File: algorithms/166.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class Solution {
public:
    string fractionToDecimal(long numerator, long denominator) {
        if (numerator == 0)
            return "0";
        string result;
        if (numerator < 0 ^ denominator < 0)
            result += "-";
        numerator = abs(numerator);
        denominator = abs(denominator);
        result += to_string(numerator/denominator);
        numerator %= denominator;
        if (numerator == 0)
            return result;
        result += ".";
        unordered_map<long, int> m;
        while (numerator) {
            if (m.find(numerator) == m.end()) {
                m[numerator] = result.size();
                numerator *= 10;
                result += to_string(numerator/denominator);
                numerator %= denominator;
            }
            else {
                result.insert(m[numerator], 1, '(');
                result.insert(result.end(), 1, ')');
                break;
            }
        }
        return result;
    }
};

int main() {
    return 0;
}

File: algorithms/python/166.py
class Solution:
    def fractionToDecimal(self, numerator, denominator):
        """
        :type numerator: int
        :type denominator: int
        :rtype: str
        """
        sign = '-' if numerator * denominator < 0 else ''
        numerator, denominator = abs(numerator), abs(denominator)
        rls = [str(numerator // denominator), '.']
        numerator %= denominator
        if numerator == 0:
            return sign + rls[0]
        m = {numerator: len(rls)}
        while numerator:
            numerator *= 10
            rls.append(str(numerator // denominator))
            numerator %= denominator
            if numerator in m:
                return sign + ''.join(rls[:m[numerator]]) + '(' + ''.join(rls[m[numerator]:]) + ')'
            m[numerator] = len(rls)
        return sign + ''.join(rls)
